init/devices.cpp

```diff
@@ -402,40 +414,96 @@ std::vector<std::string> DeviceHandler::GetBlockDeviceSymlinks(const Uevent& uev
                                  const std::string& upath, bool block,
                                  int major, int minor, const std::vector<std::string>& links) const {
-    std::vector<std::string> all_links(links);
-    std::string alias_link;
+    std::string aliases_linkpath;
+    std::stringstream slfmt;
 
-    // Add the alias link to the device to the list of links to
-    // create.
-    if (GetDeviceAlias(upath, major, minor, alias_link))
-        all_links.push_back(alias_link);
+    // Create a symlink in /dev/aliases that will point to the
+    // alias.
+    slfmt << "/dev/aliases/" << major << "_" << minor;
+    aliases_linkpath = slfmt.str();
 
     if (action == "add") {
+        std::string alias_link;
+        std::vector<std::string> all_links(links);
+
+        // Add the alias link to the device to the list of links to
+        // create.
+        if (GetDeviceAlias(upath, major, minor, alias_link))
+            all_links.push_back(alias_link);
+
         MakeDevice(devpath, block, major, minor, all_links);
+
         for (const auto& link : all_links) {
-            if (!mkdir_recursive(Dirname(link), 0755)) {
+            std::string link_path;
+
+            if (!mkdir_recursive(Dirname(link), 0755))
                 PLOG(ERROR) << "Failed to create directory " << Dirname(link);
-            }
 
-            if (symlink(devpath.c_str(), link.c_str())) {
-                if (errno != EEXIST) {
-                    PLOG(ERROR) << "Failed to symlink " << devpath << " to " << link;
-                } else if (std::string link_path;
-                           Readlink(link, &link_path) && link_path != devpath) {
-                    PLOG(ERROR) << "Failed to symlink " << devpath << " to " << link
-                                << ", which already links to: " << link_path;
+            // Only create a link in /dev/aliases for device
+            // aliases. Ordinary device symlinks will be created
+            // within the 'else'.
+            if (!alias_link.empty()
+                && link == alias_link
+                && symlink(link.c_str(), aliases_linkpath.c_str())) {
+
+                if (errno != EEXIST)
+                    PLOG(ERROR) << "Failed to create alias symlink from " << aliases_linkpath
+                                << " to " << link;
+
+                else if (Readlink(link, &link_path) && link_path != devpath)
+                    PLOG(ERROR) << "Failed to create alias symlink from " << aliases_linkpath
+                                << " to " << link << ", which already links to: " << link_path;
+            }
+            else {
+                if (symlink(devpath.c_str(), link.c_str())) {
+                    if (errno != EEXIST)
+                        PLOG(ERROR) << "Failed to symlink " << devpath
+                                    << " to " << link;
+
+                    else if (Readlink(link, &link_path) && link_path != devpath)
```

Comment:
For the error diagnostic, you're looking at `link_path` but the link you attempted to create was in `link`. Is that alright?

Reply:
I'll double check that but I'll try to come up with clearer variable names too because at this point I'm confusing myself.